root.c

```diff
@@ -1922,6 +1938,161 @@ static void notify_thread(struct unlocked_watchman_root *unlocked)
   w_pending_coll_destroy(&pending);
 }
 
+#ifndef _WIN32
+// Given a target of the form "absolute_path/filename", return
+// realpath(absolute_path) + filename, where realpath(absolute_path) resolves
+// all the symlinks in absolute_path.
+static w_string_t *get_normalized_target(w_string_t *target) {
+  w_string_t *dir_name, *file_name, *normalized_target = NULL;
+  char *dir_name_buf, *dir_name_real;
+
+  w_assert(w_is_path_absolute(target->buf),
+           "get_normalized_target: path %s is not absolute\n", target->buf);
+  dir_name = w_string_dirname(target);
+  // Need a duplicated buffer to get terminating null character
+  dir_name_buf = w_string_dup_buf(dir_name);
+  file_name = w_string_basename(target);
+  dir_name_real = w_realpath(dir_name_buf);
+  if (dir_name_real) {
+    w_string_t *dir_name_real_wstr =
+      w_string_new_typed(dir_name_real, W_STRING_BYTE);
+    normalized_target = w_string_path_cat(dir_name_real_wstr, file_name);
+    w_string_delref(dir_name_real_wstr);
+    free(dir_name_real);
+  }
+  w_string_delref(dir_name);
+  w_string_delref(file_name);
+  free(dir_name_buf);
+  return normalized_target;
+}
+
+// Requires target to be an absolute path
+static void watch_symlink_target(w_string_t *target) {
+  char *watched_root = NULL, *relpath = NULL;
+  w_string_t *normalized_target;
+
+  w_assert(w_is_path_absolute(target->buf),
+           "watch_symlink_target: path %s is not absolute\n", target->buf);
+  normalized_target = get_normalized_target(target);
+  if (!normalized_target) {
+    w_log(W_LOG_ERR, "watch_symlink_target: "
+          "failed to get normalized version of target %s\n", target->buf);
+    return;
+  }
+  watched_root = w_find_enclosing_root(normalized_target->buf, &relpath);
+  if (watched_root) {
+    // We are already watching a root that contains this target
+    free(watched_root);
+    if (relpath) {
+      free(relpath);
+    }
+  } else {
+    char *resolved, *errmsg;
+    json_t *root_files;
+    bool enforcing, res;
+    resolved = strdup(normalized_target->buf);
+    root_files = cfg_compute_root_files(&enforcing);
```

Comment:
(and by caller, I mean the function that iterates over the set of symlink candidates)